src/store/index-level.ts

```diff
@@ -0,0 +1,216 @@
+const KEY_PREFIX_DELIMITER = '@';
+
+export type Entry = {
+  _id: string,
+  [property: string]: unknown
+};
+
+export type ExactFilter = string | number | boolean;
+
+export type RangeFilter = {
+  gte?: string,
+  lte?: string
+};
+
+export type Filter = {
+  [property: string]: ExactFilter | ExactFilter[] | RangeFilter
+};
+
+export interface IndexLevelOptions {
+  signal?: AbortSignal;
+}
+
+export class IndexLevel {
+  config: IndexLevelConfig;
+
+  db: LevelDatabase<string, string>;
+
+  constructor(location: string, config?: IndexLevelConfig) {
+    this.config = {
+      createLevelDatabase,
+      ...config
+    };
+
+    this.db = this.config.createLevelDatabase<string, string>(location, { keyEncoding: 'utf8', valueEncoding: 'utf8' });
+  }
+
+  open(): Promise<void> {
+    return this.db.open();
+  }
+
+  close(): Promise<void> {
+    return this.db.close();
+  }
+
+  async put(entry: Entry, options?: IndexLevelOptions): Promise<void> {
+    options?.signal?.throwIfAborted();
+
+    entry = flat.flatten(entry);
+    const { _id } = entry;
+
+    const ops: AbstractBatchOperation<LevelDatabase<string, string>, string, string>[] = [];
+    const keyPrefixes: string[] = [];
+    for (const property in entry) {
+      if (property === '_id') {
+        continue;
+      }
+
+      const value = entry[property];
+
+      const keyPrefix = this.generateKey(property, value);
+      ops.push({ type: 'put', key: keyPrefix + VALUE_DELIMITER + _id, value: _id });
+      keyPrefixes.push(keyPrefix);
+    }
+    ops.push({ type: 'put', key: `__${_id}__keyPrefixes`, value: keyPrefixes.join(KEY_PREFIX_DELIMITER) });
+
+    await abortOr(options?.signal, this.db.batch(ops));
+  }
+
+  async query(filter: Filter, options?: IndexLevelOptions): Promise<Array<string>> {
+    options?.signal?.throwIfAborted();
+
+    let requiredMatchCount = 0;
+
+    const promises: Promise<void>[] = [];
+    const matches: { [docId: string]: number } = {};
+    function handleMatch(_key: string, value: string): void {
+      matches[value] = (matches[value] ?? 0) + 1;
+    }
+    for (const propertyName in filter) {
+      const propertyFilter = filter[propertyName];
+
+      if (typeof propertyFilter === 'object' && propertyFilter !== null) {
+        if (Array.isArray(propertyFilter)) {
+          for (const propertyValue of propertyFilter) {
+            promises.push(this.forEachTermMatch(propertyName, propertyValue, handleMatch, options));
+          }
+        } else {
+          promises.push(this.forEachRangeMatch(propertyName, propertyFilter, handleMatch, options));
+        }
+      } else {
+        promises.push(this.forEachTermMatch(propertyName, propertyFilter, handleMatch, options));
+      }
+
+      ++requiredMatchCount;
+    }
+
+    await Promise.all(promises);
+
+    const docIds: string[] = [];
+    for (const docId in matches) {
+      if (matches[docId] === requiredMatchCount) {
+        docIds.push(docId);
+      }
+    }
+    return docIds;
+  }
```

Comment:
i find this a bit hard to follow compared to:

```typescript
  async query(filter: Filter): Promise<Array<string>> {
    const matches: { [docId: string]: number } = {};
    const promises: Promise<void>[] = [];

    for (const propertyName in filter) {
      const propertyValue = filter[propertyName];

      if (typeof propertyValue === 'object' && propertyValue !== null) {
        if (Array.isArray(propertyValue)) {
          // OR query
          for (const value of propertyValue) {
            const promise = this.buildTermQuery(propertyName, value, matches);
            promises.push(promise);
          }
        } else {
          const promise = this.buildRangeQuery(propertyName, propertyValue, matches);
          promises.push(promise);
        }
      } else {
        const promise = this.buildTermQuery(propertyName, propertyValue, matches);
        promises.push(promise);
      }
    }

    const numFilters = Object.keys(filter).length;
    const docIds: string[] = [];
    await Promise.all(promises);

    for (const docId in matches) {
      if (matches[docId] === numFilters) {
        docIds.push(docId);
      }
    }

    return docIds;
  }
```

Reply:
what is it about the way i've written it that's more confusing?

IMO passing in `matches` and then suddenly having it be magically populated with things after the `await` is more confusing because i have to look in three different places to see all the ways that `matches` is modified (whereas the way i've written it is all in the same function).  TL;DR: i **_really_** dislike using out parameters in JS :(

would it be clearer if instead of having it take a `callback` it returned an array of all the matches?  thinking about this a bit more i think that'd be even more clean